### DISCUSSION TITLE
Suggestion: provide error types

### DISCUSSION THREAD

**Discussion by davidbnk** [stephenh/ts-proto#695] [Ideas]
Wonder if it would be useful for this package to provide error types? Something like this:

```
interface Metadata {
  get: (key: string) => string[];
}
export interface GRPCError {
  message: string;
  code: number;
  details: string;
  metadata: Metadata;
}

export function isGRPCError(err: unknown): err is GRPCError {
  let e = err as GRPCError;
  return (
    typeof e.message === "string" &&
    typeof e.code === "number" &&
    typeof e.details === "string" &&
    typeof e.metadata === "object" &&
    typeof e.metadata.get === "function"
  );
}
```

**Comment by stephenh** [comment]
@davidbnk hey! I would not be against that, as that does seem useful.

My only thought that is it seems like that `isGRPCError` should probably be in the native grpc package (or @improbable-eng/grpc-web for grpc-web/etc, not sure which runtime you're using) itself, as then:

a) they would be best suited to know "what is/is not a GRPC error" since they own the internal impl, and

b) then it'd be included in the library instead being re-codegen-d for every file/project that ts-proto would generate.

Maybe try getting this added to whatever runtime library that ts-proto is generating code on top of, and if they just don't want to accept/impl it, then yeah I wouldn't mind a PR that sneaks it into the ts-proto output as a helper method.

**Reply by davidbnk** [reply]
> they would be best suited to know "what is/is not a GRPC error" since they own the internal impl

Agree, that would make more sense. Thanks for the reply
